libs/core/execution_base/include/hpx/execution_base/traits/coroutine_traits.hpp

```diff
@@ -0,0 +1,220 @@
+//  Copyright (c) 2022 Shreyas Atre
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/config/coroutines_support.hpp>
+#include <hpx/config/forward.hpp>
+#include <hpx/concepts/has_member_xxx.hpp>
+#include <type_traits>
+
+#if defined(HPX_HAVE_CXX20_COROUTINES)
+
+namespace hpx { namespace execution { namespace experimental {
+
+    template <typename Promise, typename Awaiter>
+    decltype(auto) await_suspend(Awaiter&& await)
+    {
+        if constexpr (!std::is_same_v<Promise, void>)
+        {
+            return await.await_suspend(hpx::coro::coroutine_handle<Promise>{});
+        }
+        return;
```

Comment:
Should this be
```suggestion
    else
    {
        return;
    }
```
? IOW, what's the return type if `std::is_same_v<Promise, void>`?

Reply:
It should be void

Reply:
> It should be void

In this case I'm surprised it compiles as is. Or does `await.await_suspend(hpx::coro::coroutine_handle<Promise>{});` always return `void` as well?

Reply:
Thanks, this should have some tests. Also, await_suspend should always accept a coroutine handle --> [p2880](https://www.open-std.org/JTC1/SC22/wg21/docs/papers/2018/p1288r0.pdf) page 6

Reply:
Since `await_suspend` should either return `void` or `bool`, this code needs more attention. In the case that `await_suspend` returns a `bool` this will not compile as is (the different return paths would return different types). Thus the patch I suggested will be necessary.